### ISSUE TITLE
device twin time out

### ISSUE THREAD

**Issue by lucarv** [Azure/azure-iot-sdk-node#34]
# Context
- **OS and version used:** mac osx sierra and windows 10
- **Node.js version:** v6.9.4
- **npm version:** 3.10.10

# Description of the issue:
access to device twins failing all the time (running code as per example on sdk)
get the following error message
could not get twin: {"message":"subscription to twin messages timed out"}
same code has worked before, now it fails all the time 
have deleted device, deleted iot-hub and created new, no joy.

# Code sample exhibiting the issue:
as per https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-node-node-twin-getstarted

# Hub Connection String:
HostName=lucasmmaashub.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey=<SAK>

**Comment by pierreca** [comment]
@lucarv Thanks for reporting your issue. Can we try a couple of things to narrow down the issue?

First, can you install and run [iothub-diagnostics](https://github.com/Azure/iothub-diagnostics)?

```
$ npm install -g iothub-diagnostics
$ iothub-diagnostics <YOUR_CONNECTION_STRING>
```

Please feel free to post the console log (minus connection strings) of the result here.

I would also encourage you not to post connection strings - I'm editing it out of your issue but I suggest you rotate the keys (can be done in the Azure portal)

**Comment by lucarv** [comment]
ekskog-2:smsim epkloc$ iothub-diagnostics "HostName=lucasmmaashub.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey=5a2c0YajVIuM9ozwiVa5ejDXyTyHhXu/p14wCE2msUw="
2017-04-11T12:52:26.808Z - info: *******************************************
2017-04-11T12:52:26.812Z - info: * Executing the Microsoft IOT Trace tool. *
2017-04-11T12:52:26.812Z - info: *******************************************
2017-04-11T12:52:26.813Z - info:
2017-04-11T12:52:26.813Z - info: --- Executing network tests ---
2017-04-11T12:52:26.843Z - info:
2017-04-11T12:52:26.843Z - info: Starting DNS resolution for host 'www.microsoft.com'...
2017-04-11T12:52:26.987Z - info: --> Successfully resolved DNS to 23.14.9.230.
2017-04-11T12:52:26.988Z - info:
2017-04-11T12:52:26.988Z - info: Pinging IPV4 address '23.14.9.230'...
2017-04-11T12:52:27.016Z - info: --> Successfully pinged 23.14.9.230
2017-04-11T12:52:27.016Z - info:
2017-04-11T12:52:27.017Z - info: Sending https request to 'https://www.microsoft.com/'
2017-04-11T12:52:34.927Z - info: --> Completed https request
2017-04-11T12:52:34.927Z - info:
2017-04-11T12:52:34.928Z - info: --- Executing IOT Hub tests ---
2017-04-11T12:52:37.939Z - info:
2017-04-11T12:52:37.939Z - info: Starting AMQP Test...
2017-04-11T12:52:42.343Z - info: --> Successfully ran AMQP test.
2017-04-11T12:52:42.343Z - info:
2017-04-11T12:52:42.344Z - info: Starting AMQP-WS Test...
                         2017-04-11T12:52:47.911Z - info: --> Successfully ran AMQP-WS test.
2017-04-11T12:52:47.928Z - info:
2017-04-11T12:52:47.928Z - info: Starting HTTPS Test...
2017-04-11T12:52:52.528Z - info: --> Successfully ran HTTPS test.
2017-04-11T12:52:52.566Z - info:
2017-04-11T12:52:52.566Z - info: Starting Mqtt Test...
2017-04-11T12:52:56.765Z - info: --> Successfully ran Mqtt test.

Skickades från E-post<https://go.microsoft.com/fwlink/?LinkId=550986> för Windows 10

Från: Pierre Cauchois<mailto:notifications@github.com>
Skickat: Monday, April 10, 2017 8:33 PM
Till: Azure/azure-iot-sdk-node<mailto:azure-iot-sdk-node@noreply.github.com>
Kopia: Tayo Carvalhal<mailto:Tayo.Carvalhal@microsoft.com>; Mention<mailto:mention@noreply.github.com>
Ämne: Re: [Azure/azure-iot-sdk-node] device twin time out (#34)


@lucarv<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flucarv&data=02%7C01%7CTayo.Carvalhal%40microsoft.com%7Ca2dcf7a5b83a490e404408d480401be2%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636274460218084581&sdata=BdwTgy8oSFGjV0DBnGvN5I074dJ5xleSa9WtuMrXzc0%3D&reserved=0> Thanks for reporting your issue. Can we try a couple of things to narrow down the issue?

First, can you install and run iothub-diagnostics<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fiothub-diagnostics&data=02%7C01%7CTayo.Carvalhal%40microsoft.com%7Ca2dcf7a5b83a490e404408d480401be2%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636274460218084581&sdata=7BJuiIpFzZjFwZhtQAsLvdQNSAQYjiWmSrPR%2BsqtB%2FU%3D&reserved=0>?

$ npm install -g iothub-diagnostics

$ iothub-diagnostics <YOUR_CONNECTION_STRING>

Please feel free to post the console log (minus connection strings) of the result here.

I would also encourage you not to post connection strings - I'm editing it out of your issue but I suggest you rotate the keys (can be done in the Azure portal)

**Comment by pierreca** [comment]
Had to edit your connection string out again - please don't post it and you should rotate it now. 
Looks like MQTT connectivity is working so that's not the issue.

The subscription timeout in the SDK is a little aggressive, it's only one second. If you look at the code here:
https://github.com/Azure/azure-iot-sdk-node/blob/09304e4ca4a9dc0e0ff7271c1edef95294a08737/device/core/lib/twin.js#L23

That is where it's set. You could try setting it to a larger value (either in your node_modules folder or by cloning and linking this repo) and let us know if that helps. 

Also, another test to run, when you get this error, is to try and send a message to the device and see if you receive it. I'm trying to figure out if the problem is just with the twin receiver object or if it's with the client as a whole (believe it's disconnected or something along those lines)

**Comment by lucarv** [comment]
Changing the timeout value to 3000 sorted it out.
Shall I test further or just assume that the only problem was the time out?

PS
I refactor the CS Daily, I should change my glasses, i swear i read in github that you were asking for the CS 😊
DS

Skickades från E-post<https://go.microsoft.com/fwlink/?LinkId=550986> för Windows 10

Från: Pierre Cauchois<mailto:notifications@github.com>
Skickat: Tuesday, April 11, 2017 8:07 PM
Till: Azure/azure-iot-sdk-node<mailto:azure-iot-sdk-node@noreply.github.com>
Kopia: Tayo Carvalhal<mailto:Tayo.Carvalhal@microsoft.com>; Mention<mailto:mention@noreply.github.com>
Ämne: Re: [Azure/azure-iot-sdk-node] device twin time out (#34)


Had to edit your connection string out again - please don't post it and you should rotate it now.
Looks like MQTT connectivity is working so that's not the issue.

The subscription timeout in the SDK is a little aggressive, it's only one second. If you look at the code here:
https://github.com/Azure/azure-iot-sdk-node/blob/09304e4ca4a9dc0e0ff7271c1edef95294a08737/device/core/lib/twin.js#L23<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-iot-sdk-node%2Fblob%2F09304e4ca4a9dc0e0ff7271c1edef95294a08737%2Fdevice%2Fcore%2Flib%2Ftwin.js%23L23&data=02%7C01%7CTayo.Carvalhal%40microsoft.com%7C200de0b10185410dcbc908d481059496%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636275308345801299&sdata=2%2FyZH3Gqw6RjSPyR9x7rHGR1z8RMV6LYsjBq%2FOm1Nog%3D&reserved=0>

That is where it's set. You could try setting it to a larger value (either in your node_modules folder or by cloning and linking this repo) and let us know if that helps.

Also, another test to run, when you get this error, is to try and send a message to the device and see if you receive it. I'm trying to figure out if the problem is just with the twin receiver object or if it's with the client as a whole (believe it's disconnected or something along those lines)

**Comment by pierreca** [comment]
@lucarv Thanks for putting up with us and helping narrow down the issue. I'll reconvene with the rest of the team to discuss the hows and whys of this timeout (I don't think it's at the right place) and we should have a fix for that within the next few days or so.

As to your question "was it the only thing..?" I think so, but I cannot exclude the possibility of other bugs :) we do have a pretty serious gated build with a large set of end-to-end tests... this one thing is less of a unintended bug than it is a design choice we need to reconsider. 

I'll update the issue as we check in the fix and release new packages.

**Comment by pierreca** [comment]
Checked in a fix that increase the timeout to 2 minutes: https://github.com/Azure/azure-iot-sdk-node/commit/2e8a12a5a462c57d7502e36f0a20bbef7b06edfa
It should be released around the end of the week as part of our regular release cycle.

**Comment by pierreca** [comment]
has been released with azure-iot-device@1.1.10 and azure-iot-device-mqtt@1.1.10
